File: app/src/main/java/com/kenjigames/ividsmusic/data/AppDatabase.java
package com.kenjigames.ividsmusic.data;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

/**
 * Main Room database class for the application.
 * Manages the SQLite database instance and provides DAO accessors.
 */
@Database(entities = {Track.class}, version = 1, exportSchema = false)
public abstract class AppDatabase extends RoomDatabase {

    private static final String DATABASE_NAME = "ivids_music_db";
    private static volatile AppDatabase instance;

    /**
     * Exposes the Track Data Access Object (DAO).
     *
     * @return The TrackDao instance.
     */
    public abstract TrackDao trackDao();

    /**
     * Retrieves the thread-safe singleton instance of the Room database.
     * Uses double-checked locking to ensure single instance instantiation.
     *
     * @param context The Android context.
     * @return The AppDatabase singleton instance.
     */
    public static AppDatabase getInstance(final Context context) {
        if (instance == null) {
            synchronized (AppDatabase.class) {
                if (instance == null) {
                    instance = Room.databaseBuilder(
                            context.getApplicationContext(),
                            AppDatabase.class,
                            DATABASE_NAME
                    )
                    .fallbackToDestructiveMigration()
                    .build();
                }
            }
        }
        return instance;
    }
}

File: app/src/main/java/com/kenjigames/ividsmusic/data/AppDatabase.kt
package com.kenjigames.ividsmusic.data

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import com.kenjigames.ividsmusic.data.dao.DiscoveryScoreDao
import com.kenjigames.ividsmusic.data.dao.HistoryDao
import com.kenjigames.ividsmusic.data.dao.PlaylistDao
import com.kenjigames.ividsmusic.data.dao.TrackDao
import com.kenjigames.ividsmusic.data.entity.DiscoveryScoreEntity
import com.kenjigames.ividsmusic.data.entity.HistoryEntity
import com.kenjigames.ividsmusic.data.entity.PlaylistEntity
import com.kenjigames.ividsmusic.data.entity.PlaylistTrackCrossRef
import com.kenjigames.ividsmusic.data.entity.TrackEntity

/**
 * Main Room database singleton for IVIDS Music.
 * Manages SQLite persistence for tracks, history, playlists, and interest scores.
 */
@Database(
    entities = [
        TrackEntity::class,
        HistoryEntity::class,
        PlaylistEntity::class,
        PlaylistTrackCrossRef::class,
        DiscoveryScoreEntity::class
    ],
    version = 3,
    exportSchema = false
)
abstract class AppDatabase : RoomDatabase() {

    abstract fun trackDao(): TrackDao
    abstract fun historyDao(): HistoryDao
    abstract fun playlistDao(): PlaylistDao
    abstract fun discoveryScoreDao(): DiscoveryScoreDao

    companion object {
        private const val DATABASE_NAME = "ivids_music_db"

        @Volatile
        private var instance: AppDatabase? = null

        fun getInstance(context: Context): AppDatabase {
            return instance ?: synchronized(this) {
                instance ?: Room.databaseBuilder(
                    context.applicationContext,
                    AppDatabase::class.java,
                    DATABASE_NAME
                )
                .fallbackToDestructiveMigration()
                .fallbackToDestructiveMigrationOnDowngrade()
                .build()
                .also { instance = it }
            }
        }
    }
}
